### DISCUSSION TITLE
Problem in performance with spawning and despawning of entities.

### DISCUSSION THREAD

**Discussion by Tonaion02** [bevyengine/bevy#15471] [Q&A]
Hello guys.

In the project that i'm realizing, i need, for every frame to spawn and despawn a large number of entities.
To parallelize the code where i create commands to spawn and despawn entitites, i used ParallelCommands.

The apply_deferred is obviously sequential, and can be a large bottleneck for my project, where performance is really important.

I checked in unofficial documentation, and i discovered that a solution for my problem can be an exclusive system. How this can be a much better solution for my problem? With direct access to the world i can allocate space for this entities in a more efficient way?

**Comment by alice-i-cecile** [comment]
When spawning large batches of entities, use `spawn_batch`, not `ParallelCommands`. `ParallelCommands` increases per-command overhead, and is suited to computation-heavy tasks. By contrast, `spawn_batch` collects all of the work to be done and dramatically reduces the amount of computation and allocations required.

You can further reduce overhead by using an exclusive system, to avoid the indirection of commands, but `spawn_batch` is by far the most important optimization.

**Reply by Tonaion02** [reply]
So the most important optimization is use directly an exclusive system?
I used parallel commands because i decided how component is builded during a parallel iteration(i can't make sequential that part of the frame, because it can became itself a bottleneck).
So the solution can be to store in some manner this info without using parallel_commands. Use spawn_batch at the end of the frame and then create a system that in parallel initialize the components with the correct data, correct?
I don't know how precisely i can write this code, but i think that the idea can win...opinions?

**Reply by alice-i-cecile** [reply]
spawn_batch is the most important optimization, followed by using an exclusive system :) And yeah, if your component data is particularly complex to create, you can try parallelizing that.

**Reply by Tonaion02** [reply]
Ok, i think that in Bevy there isn't any sort of lock free data structure to save some information in a parallel code, am i right? In that case can't be something useful?

**Reply by SkiFire13** [reply]
There is `bevy_utils::Parallel`, implemented using the `thread_local` crate, which is also used under the hood by `ParallelCommands`. I don't think this is exposed by `Parallel`, but `thread_local::ThreadLocal` implements `IntoIterator`, so you should be able to use it with `spawn_batch`.

**Comment by SkiFire13** [comment]
Commands have an overhead due to storing such commands (which requires using memory) and then invoking them (which involves dynamic dispatch). If you're spawning lot of entities this might become a bottleneck, so directly spawning them using `World` might be better due to skipping those overheads.
